### ISSUE TITLE
python packages do not build when using an external Spack-built python

### ISSUE THREAD

**Issue by koysean** [LLNL/spack#38031]
### Steps to reproduce

@scheibelp

Python packages which use pip do not build when the provided python dependency is an external built in another Spack instance. I believe it's possible that any external python which does not have pip installed in its prefix will fail to build python packages via Spack, although I have not tested this thoroughly.

To reproduce with `py-cython`:
1. create a spack instance and build python, py-pip, py-setuptools, and py-wheel
2. create a second spack instance and configure `packages.yaml` to use the installed python packages from step 1 as externals
3. `spack install py-cython`

This will produce a build error saying that pip cannot be found. Inspecting the build environment shows that `PYTHONPATH` is not appropriately set to include the pip module:

``` console
> spack build-env py-cython | grep ^PYTHONPATH
PYTHONPATH=/path/to/spack/opt/spack/linux-sles15-x86_64/gcc-11.2.0/py-cython-0.29.33-5ms5gi6bj3bj6sfywessqyozai7rj3fj/lib/python3.10/site-packages
```

I will note that loading py-pip in the first spack instance with `spack load py-pip` does prepend the correct path to `PYTHONPATH`. The issue probably lies somewhere in a `setup_dependent_build_environment` function, either in the python build system or in the python or pip packages themselves.

Adding py-pip as an explicit dependency of py-cython does not help, since it already set as a build dependency through the python build system: https://github.com/spack/spack/blob/f745e49d9ac17fd588cd3839897af01463610cc7/lib/spack/spack/build_systems/python.py#L201-L203

It seems like this part of the python package.py should be setting `PYTHONPATH`, but I'm not sure why it's not: https://github.com/spack/spack/blob/f745e49d9ac17fd588cd3839897af01463610cc7/var/spack/repos/builtin/packages/python/package.py#L1047-L1077

### Error message

``` console
/path/to/spack/python/bin/python3.10: No module named pip
```

### Information on your system

* **Spack:** 0.21.0.dev0 (eda21cdfbac2d3028370da18e1211ec8bc759cb1)
* **Python:** 3.6.15
* **Platform:** linux-sles15-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by koysean** [comment]
Narrowed down the issue and found a quick fix, although I'm not sure if it will have unintended consequences.

The problem comes from this line in the python package: https://github.com/spack/spack/blob/f745e49d9ac17fd588cd3839897af01463610cc7/var/spack/repos/builtin/packages/python/package.py#L1064

If the expression returns `True`, the dependency's prefix gets added to the `PYTHONPATH`. In this case, `package.extends(python_spec)` returns `False` for all of the python dependencies (`py-pip`, `py-wheel`, `py-setuptools`). This is happening because `package.extendee_spec` returns `None`, and that happens when the spec is already concretized: https://github.com/spack/spack/blob/f745e49d9ac17fd588cd3839897af01463610cc7/lib/spack/spack/package_base.py#L1254-L1263

I'm not sure why `self.spec._concrete` isn't `True` when installing `py-cython` along with `py-pip` and `python` all in the same Spack instance... but in any case, when `py-pip` is an external, it's considered concretized at the installation step and that results in its prefix not being added to `PYTHONPATH` in `py-cython`'s build environment.

An easy fix is to change the expression in that if statement like so:
```
if self.spec._concrete and not self.spec.external:
    return None
```
but I don't know how this will affect other build systems or potential edge cases. From a quick test, `python`, `py-pip`, and other python packages still install fine in the first Spack instance, and I'm able to install `py-cython` in the second Spack instance with the external python packages. I've opened a draft PR with this fix: https://github.com/spack/spack/pull/38052

Another solution I just thought of is to explicitly add direct build dependencies to `PYTHONPATH` without all the `extends()` checks, since it shouldn't interfere with non-python build processes.

**Comment by alalazo** [comment]
@scheibelp I triaged this as `impact-low`, but feel free to change the label if you think differently. Triaging causes project automation to move it to the list of known bugs, that can be viewed by users.

Q: do we want to backport the fix to v0.20.1 (I assume so, but want to double check)?
